src/ioctls/vm.rs

```diff
@@ -1258,62 +1284,76 @@ mod tests {
-        target_arch = "aarch64"
-    ))]
-    fn test_register_irqfd() {
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn test_register_unregister_irqfd() {
         let kvm = Kvm::new().unwrap();
         let vm_fd = kvm.create_vm().unwrap();
         let evtfd1 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd2 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd3 = EventFd::new(EFD_NONBLOCK).unwrap();
-        if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
-            assert!(vm_fd.create_irq_chip().is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
-        }
 
-        // On aarch64, this fails because setting up the interrupt controller is mandatory before
-        // registering any IRQ.
+        assert!(vm_fd.create_irq_chip().is_ok());
+
+        assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
+        assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
+        assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
+        assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
+        // KVM irqfd doesn't report failure on this case:(
+        assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
+
+        // Duplicated eventfd registration.
         // On x86_64 this fails as the event fd was already matched with a GSI.
         assert!(vm_fd.register_irqfd(&evtfd3, 4).is_err());
         assert!(vm_fd.register_irqfd(&evtfd3, 5).is_err());
+        // KVM irqfd doesn't report failure on this case:(
+        assert!(vm_fd.unregister_irqfd(&evtfd3, 5).is_ok());
     }
 
     #[test]
-    #[cfg(any(
-        target_arch = "x86",
-        target_arch = "x86_64",
-        target_arch = "arm",
-        target_arch = "aarch64"
-    ))]
-    fn test_unregister_irqfd() {
+    #[cfg(target_arch = "aarch64")]
+    fn test_register_unregister_irqfd() {
         let kvm = Kvm::new().unwrap();
         let vm_fd = kvm.create_vm().unwrap();
         let evtfd1 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd2 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd3 = EventFd::new(EFD_NONBLOCK).unwrap();
-        if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
-            assert!(vm_fd.create_irq_chip().is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
-            assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
-            // KVM irqfd doesn't report failure on this case:(
-            assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
-        }
+
+        // Create the vGIC device.
+        let vgic_fd = create_gic_device(&vm_fd, 0);
+
+        // refer: https://github.com/firecracker-microvm/firecracker/blob/master/src/arch/src/aarch64/layout.rs
```

Comment:
In general it is better to have permanent links. If we decide to move `layout.rs` in Firecracker to another crate, that link will not work any more. I don't think we need a reference here because I believe we can just have a random (but correct value) directly written in kvm_device_attr.

Reply:
I would still say to get rid of these two lines as it doesn't make a lot of sense to define `nr_irqs` (see below comment).